### PR TITLE
06-July-23 Converted stores.php to stores.md

### PR DESCRIPTION
**Pull request by Meng-Heng**
I deleted a language-keyman class in md since it was interfering with the whole page, will it be a problem in the future?

### CODE REVIEW

File: developer/language/guide/constraints.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+following a certain context, or certain context and keystroke
+combinations.
+
+## Rejecting keystrokes {#rejecting-keystrokes}
+
+To make a keystroke produce no output, you use the [`nul`
+statement](../reference/nul). For example, we could use the following
+rule to prevent the user from typing "Q":
+
+    + "Q" > nul
```
Comment (jahorton) [inline developer/language/guide/constraints.md:19]:
````suggestion
```
+ "Q" > nul
```
````

File: developer/language/guide/constraints.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+In some cases, you might want to let the user know that they typed an
+illegal key; for this you would use the [`beep`
+statement](../reference/beep). The `beep` statement, like the `nul`
+statement, produces no output, but it plays the sound assigned to
+"Default Beep" in the Sounds Control Panel.
+
+    + "Q" > beep
```
Comment (jahorton) [inline developer/language/guide/constraints.md:27]:
````suggestion
```
+ "Q" > beep
```
````

File: developer/language/guide/constraints.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+statement is only valid as the whole output, while the `beep` statement
+can be used as just a part of the output.
+
+## Rejecting keystrokes after a context {#rejecting-keystrokes-after-a-context}
+
+In order to reject a keystroke following a certain context, you make the
+rule output just the context. Logically, you use the `context` statement
+for this:
+
+    "ba" + "d" > context
```
Comment (jahorton) [inline developer/language/guide/constraints.md:39]:
````suggestion
```
+ "ba" + "d" > context
```
````

File: developer/language/guide/constraints.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+This would have the effect of ignoring the "d": the output would be the
+context, so to the user it would appear as though they had never typed
+the "d". You can also use the `beep` statement here to aurally alert the
+user that they typed a bad word.
+
+    "ba" + "d" > context beep
```
Comment (jahorton) [inline developer/language/guide/constraints.md:46]:
````suggestion
```
+ "ba" + "d" > context beep
```
````

File: developer/language/guide/constraints.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+## Rejecting both context and keystroke {#rejecting-both-context-and-keystroke}
+
+In certain special situations, you might wish to reject both the context
+and the keystroke. You can use the `nul` statement to achieve this aim:
+
+    "eras" + "e" > nul
```
Comment (jahorton) [inline developer/language/guide/constraints.md:53]:
````suggestion
```
+ "eras" + "e" > nul
```
````

File: developer/language/guide/constraints.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+This would effectively erase the context and the key. As in both
+situations above, you can also use the `beep` statement here to alert
+the user with a warning beep:
+
+    "eras" + "e" > beep
```
Comment (jahorton) [inline developer/language/guide/constraints.md:59]:
````suggestion
```
+ "eras" + "e" > beep
```
````

File: developer/language/guide/constraints.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+this case you might put specific constraint rules in the first group of
+a keyboard, and use a `nomatch` rule to process all other (allowed)
+input with another group. The separation of legal and illegal input into
+separate groups will make the keyboard's operation clearer.
+
+## See also
+
+-   [`nul` statement](../reference/nul "nul statement")
+-   [`beep` statement](../reference/beep "beep statement")
+-   [`nomatch` rule](../reference/nomatch "nomatch rule")
```
Comment (jahorton) [inline developer/language/guide/constraints.md:77]:
```
-   [`nul` statement](../reference/nul)
-   [`beep` statement](../reference/beep)
-   [`nomatch` rule](../reference/nomatch)
```

File: developer/language/guide/stores.md [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+The most common use of stores is to reduce the number of rules required
+in a keyboard by matching multiple characters rather than a single
+character in a rule.
+
+## Creating stores 
+
+Stores are created with the [`store()` statement](../reference/store):
+
+    store(storeName) storeContents
```
Comment (jahorton) [inline developer/language/guide/stores.md:18]:
````suggestion
```
store(storeName) storeContents
```
````

File: developer/language/guide/stores.md [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+Stores have a maximum length of 4095 characters (note that non-character
+elements such as virtual keys and deadkeys take more than one character
+in the store).
+
+### Examples 
+
+    store(vowels) 'aeiouAEIOU'
+    store(vowels_lower) 'ae' U+0069 U+006F U+0075
+    store(φωνήεντα) 'αειου'
+    store(vowel_keys) [K_A] [K_E] [K_I] [K_O] [K_U]
```
Comment (jahorton) [inline developer/language/guide/stores.md:42]:
````suggestion
```
store(vowels) 'aeiouAEIOU'
store(vowels_lower) 'ae' U+0069 U+006F U+0075
store(φωνήεντα) 'αειου'
store(vowel_keys) [K_A] [K_E] [K_I] [K_O] [K_U]
```
````

File: developer/language/guide/stores.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+## Using stores in rules 
+
+Stores are most often used with the `any()` and `index()` statements.
+The `any()` statement is used in the context or key section of the rule,
+and will match any character in the specified store, remembering its
+index.
+
+    any(storeName) + ... > ...
+    ... + any(storeName) > ...
```
Comment (jahorton) [inline developer/language/guide/stores.md:52]:
````suggestion
```
    any(storeName) + ... > ...
    ... + any(storeName) > ...
```
````

File: developer/language/guide/stores.md [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+`any()` statement used to obtain the index is referenced by its
+character position in the context (and key) part of the rule. In the
+first example line, the `any()` statement is the first element
+(character or statement) in the rule, so the offset is 1. In the second
+example line, if the three dots represent one single character or
+statement, the `any()` is the second element of the rule so the offset
+is 2. Since the context part of a rule can contain several characters or
+statements, the offset for the `any()` could be greater than two.
+
+    index(storeName, offset)
```
Comment (jahorton) [inline developer/language/guide/stores.md:65]:
````suggestion
```
index(storeName, offset)
```
````

File: developer/language/guide/stores.md [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+### Examples 
+
+In the following example, the `index()` statement translates the lower
+case vowels to upper case vowels:
+
+    store(lowercase) "aeiou"
+    store(uppercase) "AEIOU"
+
+    + any(lowercase) > index(uppercase, 1)    c translate vowels to upper case
```
Comment (jahorton) [inline developer/language/guide/stores.md:79]:
````suggestion
```
store(lowercase) "aeiou"
store(uppercase) "AEIOU"

+ any(lowercase) > index(uppercase, 1)    c translate vowels to upper case
```
````

File: developer/language/guide/stores.md [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+
+A more useful example shows how a vowel followed by a circumflex can be
+translated automatically into the combined character:
+
+    store(vowel) 'aeiouAEIOU'
+    store(vowel_circum) 'âêîôûÂÊÎÔÛ'
+    any(vowel) + '^' > index(vowel_circum, 1)
```
Comment (jahorton) [inline developer/language/guide/stores.md:87]:
````suggestion
```
store(vowel) 'aeiouAEIOU'
store(vowel_circum) 'âêîôûÂÊÎÔÛ'
any(vowel) + '^' > index(vowel_circum, 1)
```
````

File: developer/language/guide/constraints.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+statement, produces no output, but it plays the sound assigned to
+"Default Beep" in the Sounds Control Panel.
+
+    + "Q" > beep
+
+One difference between the `nul` and `beep` statements is that the `nul`
+statement is only valid as the whole output, while the `beep` statement
+can be used as just a part of the output.
+
+## Rejecting keystrokes after a context {#rejecting-keystrokes-after-a-context}
```
Comment (jahorton) [inline developer/language/guide/constraints.md:33]:
```suggestion
## Rejecting keystrokes after a context
```

File: developer/language/guide/constraints.md [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    "ba" + "d" > context
+
+This would have the effect of ignoring the "d": the output would be the
+context, so to the user it would appear as though they had never typed
+the "d". You can also use the `beep` statement here to aurally alert the
+user that they typed a bad word.
+
+    "ba" + "d" > context beep
+
+## Rejecting both context and keystroke {#rejecting-both-context-and-keystroke}
```
Comment (jahorton) [inline developer/language/guide/constraints.md:48]:
```suggestion
## Rejecting both context and keystroke
```

File: developer/language/guide/constraints.md [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+    "eras" + "e" > nul
+
+This would effectively erase the context and the key. As in both
+situations above, you can also use the `beep` statement here to alert
+the user with a warning beep:
+
+    "eras" + "e" > beep
+
+## Typical usage of constraints {#typical-usage-of-constraints}
```
Comment (jahorton) [inline developer/language/guide/constraints.md:61]:
```suggestion
## Typical usage of constraints
```

File: developer/language/guide/stores.md [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+design
+tutorial](/developer/current-version/guides/develop/tutorial/step-6 "Step 6: Stores, any(), and index()").
+
+## Outputting stores 
+
+The [`outs()` statement](../reference/outs) can be used to output the
+contents of a store. This can be used in another store, or in the
+context or output of a rule.
+
+    outs(storeName)
```
Comment (jahorton) [inline developer/language/guide/stores.md:106]:
````suggestion
```
    outs(storeName)
```
````

File: developer/language/guide/stores.md [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+
+Stores that contain only a single character are usually used to define
+[named constants](constants "Named Constants"){.link}.
```
Comment (jahorton) [inline developer/language/guide/stores.md:109]:
```suggestion
[named constants](constants).
```

File: developer/language/guide/stores.md [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+named constants:
+
+    store(vowels_lower) 'aeiou'
+    store(vowels_upper) 'AEIOU'
+    store(vowels) outs(vowels_lower) outs(vowels_upper)
+
+    store(vowel_a) 'a'
+    store(vowel_alpha) 'α'
+    + outs(vowel_a) > outs(vowel_alpha)
+    + $vowel_a > $vowel_alpha
```
Comment (jahorton) [inline developer/language/guide/stores.md:123]:
````suggestion
```
store(vowels_lower) 'aeiou'
store(vowels_upper) 'AEIOU'
store(vowels) outs(vowels_lower) outs(vowels_upper)

store(vowel_a) 'a'
store(vowel_alpha) 'α'
+ outs(vowel_a) > outs(vowel_alpha)
+ $vowel_a > $vowel_alpha
```
````

File: developer/language/guide/stores.md [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+Keyman 5.0 introduced the concept of system stores that replace the
+header statements used in earlier versions of Keyman. In Keyman 9.0 and
+later, system stores are also used to reflect and update system state.
+System stores are prefixed with an ampersand (`&`).
+
+### Example 
+
+This shows the [`&name` store](../reference/name) in use:
+
+    store(&name) "My First Keyboard"
```
Comment (jahorton) [inline developer/language/guide/stores.md:145]:
````suggestion
```
store(&name) "My First Keyboard"
```
````

File: developer/language/guide/stores.md [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+## IMX definition stores 
+
+Stores can be used to define an IMX interface function to be used with
+[`call()`](../reference/call). These stores should not then be used for
+any other purpose.
+
+### Example 
+
+    store(DLLFunction) "myimx.dll:KeyEvent"
+    + any(key) > call(DLLFunction)
```
Comment (jahorton) [inline developer/language/guide/stores.md:156]:
````suggestion
```
    store(DLLFunction) "myimx.dll:KeyEvent"
    + any(key) > call(DLLFunction)
```
````

File: developer/language/guide/stores.md [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+been known in the past as "keyboard options".
+
+    store(opt_composed) '0'
+    store(vowel) 'aeiouAEIOU'
+    store(vowel_circum) 'âêîôûÂÊÎÔÛ'
+     
+    + [CTRL ALT K_1] > set(opt_composed='1')
+    + [CTRL ALT K_0] > set(opt_composed='0')
+    if(opt_composed = '0') any(vowel) + '^' > index(vowel, 2) U+0302
+    if(opt_composed = '1') any(vowel) + '^' > index(vowel_circum, 2)
```
Comment (jahorton) [inline developer/language/guide/stores.md:182]:
````suggestion
```
store(opt_composed) '0'
store(vowel) 'aeiouAEIOU'
store(vowel_circum) 'âêîôûÂÊÎÔÛ'
     
+ [CTRL ALT K_1] > set(opt_composed='1')
+ [CTRL ALT K_0] > set(opt_composed='0')
if(opt_composed = '0') any(vowel) + '^' > index(vowel, 2) U+0302
if(opt_composed = '1') any(vowel) + '^' > index(vowel_circum, 2)
```
````

File: developer/language/guide/stores.md [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+
+[`save()` statement](../reference/save "save() statement")
+
+Persists the content of a variable store
+
+[`set()` statement](../reference/set "set() statement")
+
+Updates the content of a variable store
+
+[`store()` statement](../reference/store "store() statement")
```
Comment (jahorton) [inline developer/language/guide/stores.md:223]:
```suggestion
The following statements are used with stores:

[`any()` statement](../reference/any)

Matches on an array of characters

[`call()` statement](../reference/call)

Calls an Input Method Extension

[`if()` statement](../reference/if)

Tests the content of a variable store

[`index()` statement](../reference/_index)

Outputs from an array of characters

[`outs()` statement](../reference/outs)

Outputs an array of characters

[`reset()` statement](../reference/reset)

Resets the content of a variable store to the default

[`save()` statement](../reference/save)

Persists the content of a variable store

[`set()` statement](../reference/set)

Updates the content of a variable store

[`store()` statement](../reference/store)
```

File: developer/language/guide/constraints.md
Comment (jahorton) [inline developer/language/guide/constraints.md:1]: Note:  this file is also changed by #739.  Make any changes there and wait for a merge first before making them here, as the changes suggested in this file should already be covered there.

File: developer/language/guide/stores.md [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+sequence of Unicode character constants.
+
+All stores must be given unique identifiers, and are visible throughout
+the keyboard file. A `store()` statement can be placed anywhere in a
+keyboard source file, and its visibility is unaffected: it can even be
+placed after references to the store in other statements.
+
+Apart from characters, two other elements are allowed in stores:
+[deadkeys](../reference/deadkey "deadkey statement") and [virtual
+keys](virtual-keys "Virtual Keys and Virtual Character Keys").
```
Comment (jahorton) [inline developer/language/guide/stores.md:31]:
```suggestion
Apart from characters, two other elements are allowed in stores:
[deadkeys](../reference/deadkey) and [virtual keys](virtual-keys).
```

**Review by jahorton — CHANGES_REQUESTED** [review]
Apologies for all the requested changes; almost all of these will be covered by the updates I've made to the conversion tool (now hosted at https://github.com/jahorton/keyman-site-page-converter).

**Review by jahorton — APPROVED** [review]

**Comment by mcdurdin** [comment]
> I deleted a language-keyman class in md since it was interfering with the whole page, will it be a problem in the future?

If I understand you, it was something like this left in the Markdown file:

```
<pre class="language-keyman"><code>
```

Yes, it is correct to remove that -- pandoc should have removed that automatically!

**Comment by darcywong00** [comment]
This is merged into staging. I see the CI link checker on other PR's are also flagging references to `stores.php` (should change to `stores`)
* /developer/language/guide/headers.md
* /developer/language/reference/_index.php

**Comment by jahorton** [comment]
> This is merged into staging. I see the CI link checker on other PR's are also flagging references to `stores.php` (should change to `stores`)
> 
>     * /developer/language/guide/headers.md
> 
>     * /developer/language/reference/_index.php

Did you figure out which files were the issue manually?  Or is that available somewhere buried within the link-checker report?  All I see is stuff _implying_ that the visited page is a 404-page, but not even _that_... explicitly, at least.

**Comment by darcywong00** [comment]
Sorry the link checker didn't show anything helpful, so I just locally searched my folder for the php files

**Pull request merged**